File: apps/apple-native/Sources/ACoolProfile/ACoolProfile.swift
import Foundation

enum ACoolProfileType: String, Codable, CaseIterable, Sendable {
    case collector
    case dealer
    case vendor
    case breaker
    case consignor
    case shop
    case submissionCenter = "submission_center"
    case eventOrganizer = "event_organizer"
    case contentCreator = "content_creator"
    case administrator
}

enum ACoolVisibility: String, Codable, CaseIterable, Sendable {
    case `private`
    case connections
    case members
    case `public`
}

enum ACoolColorVisionMode: String, Codable, CaseIterable, Sendable {
    case standard
    case protanopia
    case deuteranopia
    case tritanopia
    case monochrome
}

enum ACoolNativePlatform: String, Codable, CaseIterable, Sendable {
    case android
    case wearOS = "wear_os"
    case androidXR = "android_xr"
    case metaQuest = "meta_quest"
    case iOS = "ios"
    case iPadOS = "ipados"
    case macOS = "macos"
    case watchOS = "watchos"
    case visionOS = "visionos"
    case web
}

struct ACoolPrivacy: Codable, Hashable, Sendable {
    var profileVisibility: ACoolVisibility = .private
    var collectionVisibility: ACoolVisibility = .private
    var valueVisibility: ACoolVisibility = .private
    var wishlistVisibility: ACoolVisibility = .private
    var eventAttendanceVisibility: ACoolVisibility = .private
    var allowSearchIndexing = false
    var allowProfileRecommendations = true
    var allowResearchAnalytics = false
}

struct ACoolAccessibility: Codable, Hashable, Sendable {
    var textScale: Double = 1.0
    var reduceMotion = false
    var highContrast = false
    var screenReaderOptimized = false
    var colorVisionMode: ACoolColorVisionMode = .standard
    var hapticsEnabled = true
    var speechRate: Double = 1.0
    var captionsEnabled = true

    func validated() throws -> Self {
        guard (0.8...2.0).contains(textScale) else {
            throw ValidationError.invalidTextScale
        }
        guard (0.5...2.0).contains(speechRate) else {
            throw ValidationError.invalidSpeechRate
        }
        return self
    }

    enum ValidationError: Error {
        case invalidTextScale
        case invalidSpeechRate
    }
}

struct ACoolInterests: Codable, Hashable, Sendable {
    var categories: Set<String> = []
    var franchises: Set<String> = []
    var sports: Set<String> = []
    var games: Set<String> = []
    var players: Set<String> = []
    var characters: Set<String> = []
    var teams: Set<String> = []
    var sets: Set<String> = []
    var artists: Set<String> = []
    var eras: Set<String> = []
    var cardTypes: Set<String> = []

    var allDisplayValues: [String] {
        Array(
            franchises
                .union(sports)
                .union(games)
                .union(players)
                .union(characters)
                .union(teams)
                .union(sets)
        ).sorted()
    }
}

struct ACoolTrust: Codable, Hashable, Sendable {
    var emailVerified: Bool
    var phoneVerified = false
    var mfaEnrolled: Bool
    var passkeyCount: Int
    var businessVerificationStatus: String
    var accountStanding = "good"

    var requiresStepUpAuthentication: Bool {
        !mfaEnrolled || passkeyCount == 0 || accountStanding != "good"
    }
}

struct ACoolDeviceLink: Codable, Hashable, Identifiable, Sendable {
    let id: UUID
    var platform: ACoolNativePlatform
    var trusted: Bool
    var lastSeenAt: Date
    var pushEnabled = false

    enum CodingKeys: String, CodingKey {
        case id = "deviceId"
        case platform
        case trusted
        case lastSeenAt
        case pushEnabled
    }
}

struct ACoolProfile: Codable, Hashable, Identifiable, Sendable {
    let id: UUID
    let userId: UUID
    var organizationId: UUID?
    var username: String
    var displayName: String?
    var bio: String?
    var avatarAssetId: UUID?
    var homeRegion: String?
    var preferredCurrency = "USD"
    var profileTypes: Set<ACoolProfileType>
    var interests = ACoolInterests()
    var privacy = ACoolPrivacy()
    var accessibility = ACoolAccessibility()
    var trust: ACoolTrust
    var deviceLinks: [ACoolDeviceLink] = []
    let createdAt: Date
    var updatedAt: Date
    var schemaVersion = 1

    enum CodingKeys: String, CodingKey {
        case id = "profileId"
        case userId
        case organizationId
        case username
        case displayName
        case bio
        case avatarAssetId
        case homeRegion
        case preferredCurrency
        case profileTypes = "profileType"
        case interests
        case privacy
        case accessibility
        case trust
        case deviceLinks
        case createdAt
        case updatedAt
        case schemaVersion
    }

    func validated() throws -> Self {
        let usernameExpression = try NSRegularExpression(pattern: "^[A-Za-z0-9._-]{3,40}$")
        let range = NSRange(username.startIndex..<username.endIndex, in: username)
        guard usernameExpression.firstMatch(in: username, range: range) != nil else {
            throw ValidationError.invalidUsername
        }
        guard displayName?.count ?? 0 <= 100 else {
            throw ValidationError.displayNameTooLong
        }
        guard bio?.count ?? 0 <= 500 else {
            throw ValidationError.bioTooLong
        }
        guard preferredCurrency.range(of: "^[A-Z]{3}$", options: .regularExpression) != nil else {
            throw ValidationError.invalidCurrency
        }
        guard !profileTypes.isEmpty else {
            throw ValidationError.missingProfileType
        }
        _ = try accessibility.validated()
        return self
    }

    enum ValidationError: Error {
        case invalidUsername
        case displayNameTooLong
        case bioTooLong
        case invalidCurrency
        case missingProfileType
    }
}

File: apps/android-native/src/main/kotlin/com/acoolcollector/nativeapp/profile/ACoolProfile.kt
package com.acoolcollector.nativeapp.profile

import java.time.Instant
import java.util.UUID

enum class ProfileType {
    COLLECTOR,
    DEALER,
    VENDOR,
    BREAKER,
    CONSIGNOR,
    SHOP,
    SUBMISSION_CENTER,
    EVENT_ORGANIZER,
    CONTENT_CREATOR,
    ADMINISTRATOR
}

enum class Visibility {
    PRIVATE,
    CONNECTIONS,
    MEMBERS,
    PUBLIC
}

enum class ColorVisionMode {
    STANDARD,
    PROTANOPIA,
    DEUTERANOPIA,
    TRITANOPIA,
    MONOCHROME
}

enum class NativePlatform {
    ANDROID,
    WEAR_OS,
    ANDROID_XR,
    META_QUEST,
    IOS,
    IPADOS,
    MACOS,
    WATCHOS,
    VISIONOS,
    WEB
}

data class ACoolPrivacy(
    val profileVisibility: Visibility = Visibility.PRIVATE,
    val collectionVisibility: Visibility = Visibility.PRIVATE,
    val valueVisibility: Visibility = Visibility.PRIVATE,
    val wishlistVisibility: Visibility = Visibility.PRIVATE,
    val eventAttendanceVisibility: Visibility = Visibility.PRIVATE,
    val allowSearchIndexing: Boolean = false,
    val allowProfileRecommendations: Boolean = true,
    val allowResearchAnalytics: Boolean = false
)

data class ACoolAccessibility(
    val textScale: Float = 1.0f,
    val reduceMotion: Boolean = false,
    val highContrast: Boolean = false,
    val screenReaderOptimized: Boolean = false,
    val colorVisionMode: ColorVisionMode = ColorVisionMode.STANDARD,
    val hapticsEnabled: Boolean = true,
    val speechRate: Float = 1.0f,
    val captionsEnabled: Boolean = true
) {
    init {
        require(textScale in 0.8f..2.0f) { "textScale must be between 0.8 and 2.0" }
        require(speechRate in 0.5f..2.0f) { "speechRate must be between 0.5 and 2.0" }
    }
}

data class ACoolInterests(
    val categories: Set<String> = emptySet(),
    val franchises: Set<String> = emptySet(),
    val sports: Set<String> = emptySet(),
    val games: Set<String> = emptySet(),
    val players: Set<String> = emptySet(),
    val characters: Set<String> = emptySet(),
    val teams: Set<String> = emptySet(),
    val sets: Set<String> = emptySet(),
    val artists: Set<String> = emptySet(),
    val eras: Set<String> = emptySet(),
    val cardTypes: Set<String> = emptySet()
)

data class ACoolTrust(
    val emailVerified: Boolean,
    val phoneVerified: Boolean = false,
    val mfaEnrolled: Boolean,
    val passkeyCount: Int,
    val businessVerificationStatus: String,
    val accountStanding: String = "good"
) {
    init {
        require(passkeyCount >= 0) { "passkeyCount cannot be negative" }
    }
}

data class ACoolDeviceLink(
    val deviceId: UUID,
    val platform: NativePlatform,
    val trusted: Boolean,
    val lastSeenAt: Instant,
    val pushEnabled: Boolean = false
)

data class ACoolProfile(
    val profileId: UUID,
    val userId: UUID,
    val organizationId: UUID? = null,
    val username: String,
    val displayName: String? = null,
    val bio: String? = null,
    val avatarAssetId: UUID? = null,
    val homeRegion: String? = null,
    val preferredCurrency: String = "USD",
    val profileTypes: Set<ProfileType>,
    val interests: ACoolInterests = ACoolInterests(),
    val privacy: ACoolPrivacy = ACoolPrivacy(),
    val accessibility: ACoolAccessibility = ACoolAccessibility(),
    val trust: ACoolTrust,
    val deviceLinks: List<ACoolDeviceLink> = emptyList(),
    val createdAt: Instant,
    val updatedAt: Instant,
    val schemaVersion: Int = 1
) {
    init {
        require(username.matches(Regex("^[A-Za-z0-9._-]{3,40}$"))) {
            "username must be 3-40 characters and contain only letters, numbers, dot, underscore, or hyphen"
        }
        require(displayName == null || displayName.length <= 100)
        require(bio == null || bio.length <= 500)
        require(preferredCurrency.matches(Regex("^[A-Z]{3}$")))
        require(profileTypes.isNotEmpty())
        require(schemaVersion >= 1)
    }

    fun canExposeCollectionValue(): Boolean = privacy.valueVisibility == Visibility.PUBLIC

    fun requiresStepUpAuthentication(): Boolean =
        !trust.mfaEnrolled || trust.passkeyCount == 0 || trust.accountStanding != "good"
}
